velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp

```diff
@@ -199,5 +199,120 @@ TEST_F(AverageAggregationTest, partialResults) {
+       makeFlatVector<int32_t>(
+           batchSize, [](auto row) { return row % 5; }, nullEvery(8)),
+       vec});
+  createDuckDbTable({data});
+
+  auto planNode =
+      PlanBuilder()
+          .values({data})
+          .project({"c0", "c2"})
+          .addNode([&](std::string id, core::PlanNodePtr input) {
+            core::AggregationNode::Step aggStep =
+                core::AggregationNode::Step::kPartial;
+            std::vector<core::FieldAccessTypedExprPtr> veloxGroupingExprs;
+            veloxGroupingExprs.emplace_back(
+                std::make_shared<core::FieldAccessTypedExpr>(INTEGER(), "c0"));
+            std::vector<std::shared_ptr<const core::FieldAccessTypedExpr>>
+                aggregateMasks(1);
+            std::vector<std::shared_ptr<const core::FieldAccessTypedExpr>>
+                preGroupingExprs = {};
+            std::vector<std::string> aggOutNames = {"a1"};
+            std::vector<std::shared_ptr<const core::CallTypedExpr>> aggExprs;
+            std::vector<std::shared_ptr<const core::ITypedExpr>> aggParams;
+            aggParams.emplace_back(
+                std::make_shared<core::FieldAccessTypedExpr>(DOUBLE(), "c2"));
+            aggExprs.emplace_back(std::make_shared<const core::CallTypedExpr>(
+                ROW({"col_0", "col_1"}, {DOUBLE(), BIGINT()}),
+                std::move(aggParams),
+                "avg"));
+
+            return std::make_shared<core::AggregationNode>(
+                id,
+                aggStep,
+                veloxGroupingExprs,
+                preGroupingExprs,
+                aggOutNames,
+                aggExprs,
+                aggregateMasks,
+                false,
+                input);
+          })
+          .project({"c0", "a1.col_0", "a1.col_1"})
+          .planNode();
+
+  assertQuery(planNode, "SELECT c0, sum(c2), count(c2) FROM tmp group by c0");
+}
+
+TEST_F(AverageAggregationTest, finalAvgForNullsWithRowConstructor) {
+  auto batchSize = 1'000;
+  auto rowType = ROW({"c0", "c1", "c2"}, {INTEGER(), DOUBLE(), BIGINT()});
+
+  auto sumVec = makeFlatVector<double>(batchSize, [](auto row) { return row; });
+  for (auto i = 0; i < sumVec->size() - 1; i++) {
+    // One value is not null to avoid constant mapping.
+    sumVec->setNull(i, true);
+  }
+  auto countVec =
+      makeFlatVector<int64_t>(batchSize, [](auto row) { return 0; });
+
+  auto data = makeRowVector(
+      {makeFlatVector<int32_t>(
+           batchSize, [](auto row) { return row % 100; }, nullEvery(6)),
+       sumVec,
+       countVec});
+  createDuckDbTable({data});
+
+  auto planNode = PlanBuilder()
+                      .values({data})
+                      .project({"c0", "row_constructor(c1, c2) as p0"})
+                      .finalAggregation({"c0"}, {"avg(p0)"}, {DOUBLE()})
+                      .planNode();
+
+  assertQuery(planNode, "SELECT c0, sum(c1) / count(c2) FROM tmp group by c0");
+}
+
+TEST_F(AverageAggregationTest, fullAvgForNulls) {
```

Comment:
@rui-mo I tried running this test on the code from 'main' branch and it is passing. Is this expected? I'd assume it would fail and illustrate a bug.

Reply:
Yes, this test works, and it was used for a  comparison. If the row(sum, count) does not lose the validity information, the nulls can be correctly handled. But if the row(sum, count) comes from a row_constructor, the validity information is not there, as shown in `TEST_F(AverageAggregationTest, finalAvgForNullsWithRowConstructor)`, nulls will be accessed like valid values.